### DISCUSSION TITLE
Confused about the server/client stuff

### DISCUSSION THREAD

**Discussion by Miq1** [eModbus/eModbus#214] [General]
@emign: I moved this to Discussions, since I believe it may better fit there.

Original question was:

Hey, sorry to hijack this, but I keep getting confused with the modbus server/client stuff.

Let me explain what I want to do. Maybe you can help me figure out if the solution described in this thread is something for me:

Data source is a controller for an electric wall charger (with modbus 485).  Register table is here: [Modbus RS485.pdf](https://github.com/eModbus/eModbus/files/9331114/Modbus.RS485.pdf). It is supported by this EV charge controller: [evcc](https://github.com/evcc-io/evcc). 
But I need to add it to the software via modbus over LAN/Wifi. It has to read **and** write registers

Is this bridge (RTU client => ESP32 BRIDGE => TCP server) correct for this use case? What confuses me is, that this snippet seems to "just" send to a modbus TCP server and as how I understand it, evcc is not a server but polls the modbus device for information or sends modbus write commands to it.

_Originally posted by @emign in https://github.com/eModbus/eModbus/discussions/211#discussioncomment-3388788_

**Comment by Miq1** [comment]
As a rule of thumb, Modbus normally cannot mix both RTU (RS485) and TCP communications. If you need to connect one participant in RTU with another in TCP, you will need a bridge. A bridge is always directional, as it will act as a server in one domain and as a client in the other.

So if you need to access a server on a RTU bus, the bridge needs to be the one and only client on that bus. It will be a TCP bridge, acting as a TCP server for other TCP clients requesting.

If one TCP client needs to read or modify registers of a server on the RTU bus, it will address these requests to the bridge, that needs to be configured to forward these requests to the RTU server proper and respond with the response it got from thar RTU server to the requesting TCP client in turn.

Same holds true for a RTU bridge, that will act as a RTU server on the bus, but forward the requests of the single RTU client on the bus to attached TCP servers and bring back their responses.

**Reply by emign** [reply]
Thanks. That clears up some things. Is there actually a code example for this somewhere here ?

RTU Server <=> ESP32(RTU Client/TCP Server) <=> TCP client

**Reply by Miq1** [reply]
There is a [WiFi-RTU bridge](https://github.com/eModbus/eModbus/blob/master/examples/WiFi-RTU_Bridge/main.cpp) among the examples. As you have read above, this is a TCP server forwarding requests to a RTU bus.

**Reply by emign** [reply]
Thank you very much.

**Reply by Miq1** [reply]
Off the record: hallo Nicolai, zur Not können wir auch auf Deutsch kommunizieren. 😉

**Reply by emign** [reply]
> Off the record: hallo Nicolai, zur Not können wir auch auf Deutsch kommunizieren. 😉

Meinst du, es wäre möglich, dass du mir noch ein bisschen hilfst? Ich habe jetzt eigentlich alles soweit aufgesetzt und angeschlossen, erhalte aber keine Modbus Antwort und bin mit meinem Latein am Ende

**Reply by Miq1** [reply]
Na sicher!

@all: for the sake of precision we will continue in German here - sorry. Use DeepL or another translator to keep track if you like.

@emign: hast du mal deinen Sketch und ein Bild deiner Verdrahtung für mich?

**Reply by bertmelis** [reply]
FYI: Ich konnte auch auf Deutsch reden. Nur, es ist viele Jahre her, dass ich einen grundlegenden Deutschunterricht hatte.

**Reply by emign** [reply]
I translated :)

**Reply by bertmelis** [reply]
Please, write in German. I'll learn something. I will keep replying in English though.

**Comment by emign** [comment]
I just copy/pasted the [RTU Bridge](https://github.com/eModbus/eModbus/blob/master/examples/WiFi-RTU_Bridge/main.cpp). The serial port is set to 9600 Baud tho. The ESP gets a Wifi IP Adress but I cannot connect to the modbus

This is the [MAX3485](https://www.amazon.de/gp/product/B096ZXXKCR/ref=ppx_yo_dt_b_asin_title_o02_s00?ie=UTF8&psc=1):
<img width="500" alt="image" src="https://user-images.githubusercontent.com/5278352/184858484-eaeda156-0241-48f6-9010-d8216633ac0a.png">
At D+/A and D-/B I connected the corresponding EVSE Modbus ports. GND on modbus side is not connected

Thats the ESP32 (Blue cable to 3V3)
<img width="500" alt="image" src="https://user-images.githubusercontent.com/5278352/184858636-30d26e74-ee15-4aa8-a1a3-7df1301985d3.png">

**Reply by bertmelis** [reply]
Does the board have a 120 Ohms termination? What is the (idle) voltage between A and B? Does swapping the wires do anything?

EDIT: you do have a 120 termination, but only if you short circuit R16. alternatively, just screw in a 120 resistor into the terminals between the A and B.
On short wires you probably won't suffer from reflection but it surely puts a stable idle voltage on the bus.

**Reply by Miq1** [reply]
@bertmelis I will stick to German unless you will shout at me 😉 

@emign Hast du einen Logikanalysator oder ein Oszi verfügbar? Es ist oft hilfreich gewesen, das Signal direkt an den Pins zu sehen und erst danach in den Code einzusteigen.
Eine Skizze mit allen Komponenten und ihrer Verdrahtung hilft auch.

**Reply by emign** [reply]
The voltage between a and b is 3V (2.93)
I do not have an oszi :/

**Reply by bertmelis** [reply]
Add a 120 resistor between A and B, or short circuit R16, but that seems quite definite.

**Reply by Miq1** [reply]
3V ist viel bei 3V3 Basisspannung. Hast du an *einem* Ende des Modbus einen Pullup von A nach Vcc und einen Pulldown von B nach GND? Ist bei allen Busteilnehmern A mit A verbunden und B mit B und nichts vertauscht?

**Reply by emign** [reply]
> 3V ist viel bei 3V3 Basisspannung. Hast du an *einem* Ende des Modbus einen Pullup von A nach Vcc und einen Pulldown von B nach GND? Ist bei allen Busteilnehmern A mit A verbunden und B mit B und nichts vertauscht?

Das ist die idle Spannung ohne Bus Teilnehmer

**Reply by Miq1** [reply]
Wenn du statt der Bridge einen RTU-Client auf den ESP32 lädst - kannst du dann den Server abfragen (Log sollte als Nachweis reichen)?

**Reply by emign** [reply]
> Wenn du statt der Bridge einen RTU-Client auf den ESP32 lädst - kannst du dann den Server abfragen (Log sollte als Nachweis reichen)?

Das ist eine gute Idee. Werde ich probieren und melde mich dann.

**Reply by Miq1** [reply]
Okay, ich bin gleich sowieso 2-3 Stunden unterwegs - kannst dir also Zeit lassen.

**Reply by Miq1** [reply]
Bin wieder da! 👋🏻

**Reply by emign** [reply]
So. Ich habe jetzt nochmal nachgeforscht und gesehen, dass es genau für meine Steuerung vom Hersteller eine esp8266 Firmware mit modbus gibt die ich nehmen kann, ohne etwas selbst Stricken zu müssen

**Reply by Miq1** [reply]
Okay - was immer dir am besten erscheint! Viel Erfolg!
